src/include/platform/OTARequestorInterface.h

```diff
@@ -30,6 +30,115 @@
+         */
+        Iterator(const Optional<app::Clusters::OtaSoftwareUpdateRequestor::Structs::ProviderLocation::Type> * const list,
+                 size_t size)
+        {
+            mList         = list;
+            mListSize     = size;
+            mCurrentIndex = 0;
+            mNextIndex    = 0;
+        }
+
+        /**
+         * Search for the next provider location found in the list
+         */
+        bool Next()
+        {
+            for (size_t i = mNextIndex; i < mListSize; i++)
+            {
+                if (mList[i].HasValue())
+                {
+                    mCurrentIndex = i;
+                    mNextIndex    = mCurrentIndex + 1;
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        /**
+         * Retrieves a reference to the provider location found on a previous call to Next()
+         */
+        const app::Clusters::OtaSoftwareUpdateRequestor::Structs::ProviderLocation::Type & GetValue() const
+        {
+            return mList[mCurrentIndex].Value();
+        }
+
+    private:
+        const Optional<app::Clusters::OtaSoftwareUpdateRequestor::Structs::ProviderLocation::Type> * mList;
+        size_t mListSize;
+        size_t mCurrentIndex;
+        size_t mNextIndex;
+    };
+
+    /**
+     * Retrieve an iterator to the list
+     */
+    Iterator Begin() const { return Iterator(mList, CHIP_CONFIG_MAX_FABRICS); }
+
+    /**
+     * Add a provider location to the list if there is space available
+     */
+    CHIP_ERROR Add(const app::Clusters::OtaSoftwareUpdateRequestor::Structs::ProviderLocation::Type & providerLocation)
+    {
+        for (size_t i = 0; i < mMaxSize; i++)
+        {
+            if (!mList[i].HasValue())
+            {
+                mList[i].SetValue(providerLocation);
+                mListSize++;
+                return CHIP_NO_ERROR;
+            }
+        }
+
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    /**
+     * Delete a provider location from the list if found
+     */
+    CHIP_ERROR Delete(const app::Clusters::OtaSoftwareUpdateRequestor::Structs::ProviderLocation::Type & providerLocation)
+    {
+        for (size_t i = 0; i < mMaxSize; i++)
+        {
+            if (mList[i].HasValue())
+            {
+                app::Clusters::OtaSoftwareUpdateRequestor::Structs::ProviderLocation::Type pl = mList[i].Value();
+                if ((pl.GetFabricIndex() == providerLocation.GetFabricIndex()) &&
+                    (pl.providerNodeID == providerLocation.providerNodeID) && (pl.endpoint == providerLocation.endpoint))
+                {
+                    mList[i].ClearValue();
+                    mListSize--;
+                    return CHIP_NO_ERROR;
+                }
+            }
+        }
+
+        return CHIP_ERROR_NOT_FOUND;
+    }
+
+private:
+    Optional<app::Clusters::OtaSoftwareUpdateRequestor::Structs::ProviderLocation::Type> mList[CHIP_CONFIG_MAX_FABRICS];
```

Comment:
Do we need to use `Optional`? Given that we don't care about the order we could assume that the first `mListSize` elements are occupied. When removing i-th element we could just swap it with the last element and `mListSize--`. That way no dedicated iterator would be needed and `Span<app::Clusters::OtaSoftwareUpdateRequestor::Structs::ProviderLocation::Type>` could be returned. Thoughts?

Reply:
So the iterator is needed because the driver will need access to the list when it is determining which provider to use for the next periodic query. I'd rather expose a way to traverse rather than a handle to the list directly. With that said, the Optional is not necessary if we stores the way you mentioned (shifting as we remove). However, is there a reason you don't want to use Optional if the iterator needs to stay?

Reply:
I just noticed the memory growth and thought whether that could be avoided, but I don't have other reasons.

Reply:
I believe the majority of the growth stems from the Iterator class which is needed.